### ISSUE TITLE
Lua more vim.iter / vim.func functions

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#23233]
### Problem

`vim.iter()` was added in https://github.com/neovim/neovim/pull/23029 . 

This issue tracks some known related functions that we want in `vim.iter` or `vim.func`.

### Expected behavior

from https://github.com/neovim/neovim/pull/8677#issuecomment-403737119 : 

- (transparent) coroutine support https://www.lua.org/pil/9.3.html
- "once/memoize"
- partial application (currying? but see also the caveats mentioned [here](https://github.com/moriyalb/lamda#notice)) 
- reduce
- table management:
    - [clone](https://github.com/vigemus/nvimux/blob/master/lua/fns.lua#L7)
    - [keys](https://github.com/vigemus/iron.nvim/blob/master/lua/iron/util/functional.lua#L12)
    - [values](https://github.com/vigemus/iron.nvim/blob/master/lua/iron/util/functional.lua#L20)
    - [get](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/utils.lua#L16) (as a safe alternative to direct referencing)
    - [get-in](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/utils.lua#L20) (get in nested tables)
- functional:
    - [curry](https://github.com/vigemus/iron.nvim/blob/eac8f6330/lua/iron/utils.lua#L25)(this is an applied example)

**Comment by theHamsta** [comment]
`it1:chain(it2)` would be nice.

**Comment by stasjok** [comment]
Maybe expose also ListIter.new? So that the programmer can bypass list check loop if he knows beforehand that his table is definitely a list.
https://github.com/neovim/neovim/blob/558e4191f049662065984677a9c7cdbb1518092a/runtime/lua/vim/iter.lua#L781-L789

**Comment by justinmk** [comment]
> expose also ListIter.new? So that the programmer can bypass list check loop if he knows beforehand that his table is definitely a list.

Could do this as kwargs (`opt`) parameter on `vim.iter()` itself. In general, we should probably use that pattern for all the cases where a function accepts a table/list/etc (while keep the common case low-verbosity).

**Comment by stasjok** [comment]
I now noticed that this loop creates a copy of the table. So there probably was a reason for this.

**Comment by zjykzk** [comment]
`range(i, j)` is fine

**Comment by saccarosium** [comment]
A way to sort values would be really nice, something like `Iter:sort(<predicate>)`. Currently if you need to sort the sequence you need either sort it beforehand or first call `totable` into another variable, sort that and then recreate the iterator.

This can be a reason to go with #28941 since I'm not sure if it is something possible to do with a homogeneous interface both for lists and tables.

**Comment by ruuzia** [comment]
`Iter:count()` equivalent to `:enumerate():last()[1]`?

**Comment by justinmk** [comment]
> `Iter:count()` equivalent to `:enumerate():last()[1]`?

No, `:last()` drains the pipeline, not to mention `:enumerate()` scans all items...

**Comment by ruuzia** [comment]
> > `Iter:count()` equivalent to `:enumerate():last()[1]`?
> 
> No, `:last()` drains the pipeline, not to mention `:enumerate()` scans all items...

I must have given the wrong idea. `Iter:count()` should not be implemented as `:enumerate():last()[1]`. But yes, it should drain the pipeline and return the number of items. It's in the same simple category as `Iter:last()`, `Iter:all()`, and `Iter:any()`. Here is an example.
```lua
local num_even = vim.iter({1, 2, 3, 4, 5, 6, 7}):filter(function(x)
    return x % 2 == 0
end):count()
```
